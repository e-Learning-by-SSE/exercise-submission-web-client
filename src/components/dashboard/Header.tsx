/* Site header in Semantic ui with Login */

import React, { Component } from "react";
import { Menu, Icon, Button, Checkbox, Label, Message } from "semantic-ui-react";
import { Link } from "react-router-dom";
import DataService from "../../services/DataService";
import Account from "../account/Account";

/* React mouse event onclick */



export default class Header extends Component<React.PropsWithChildren<{onChangeVisbility: () => void, 
    onChangeDarkMode: (darkmode: boolean) => void, darkmode: boolean}>, {rightSide: any, darkmode: boolean, message: any}> {
    
    constructor(props: React.PropsWithChildren<{onChangeVisbility: () => void, onChangeDarkMode: (darkmode: boolean) => void, darkmode: boolean}>) {
        super(props);
        this.state = {rightSide: this.getLoginButton(true),
            darkmode: this.props.darkmode
            , message: null};
        this.checkIfLoggedIn();
    }

    private checkIfLoggedIn() {
        let api = new DataService();
        api.getCurrentUserDto().then((user) => {
            if (user) {
                this.setState({rightSide: <Account user={user}/>});
            } else {
                this.setState({rightSide: this.getLoginButton(false)});
            }
        }).catch((error) => {
            this.setState({rightSide: this.getLoginButton(false)});
            this.setState({message:   <Message negative size="mini" floating={true} 
                icon='warning sign'
                header='Can`t connect to server'
                content='Check your internet connection'
              />});
        });

    }

    private getLoginButton(init: boolean) {
        let localDarkmode;
        if(init) {    
            localDarkmode = false;
        } else {
            localDarkmode = this.state.darkmode;
        }
        
        
        return <Button as={Link} to="/login" inverted={localDarkmode}>Login</Button>;
    }

    private getDarkmodeCheckbox() {

        return <Checkbox toggle checked={this.state.darkmode} onChange={() => this.setDarkmode(!this.state.darkmode)}
         label={<Label>{this.state.darkmode ? "Darkmode" : "Normalmode"} <Icon inverted={this.state.darkmode} name={this.state.darkmode ? "moon" : "sun"} /></Label>}  />;
    }

    private setDarkmode(darkmode: boolean) {
        this.setState({darkmode: darkmode});
        this.props.onChangeDarkMode(darkmode);
    }

    
     handleItemClick = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
        e.preventDefault();
        this.props.onChangeVisbility();
        
       };


    render() {
        return (
        <Menu secondary inverted={this.props.darkmode}>
            <Menu.Item>
                <Button icon onClick={this.handleItemClick}>
                    <Icon inverted={this.state.darkmode} name='bars' />
                </Button>
            </Menu.Item>
            <Menu.Item position="right">
                {this.state.message}
                {this.getDarkmodeCheckbox()}
                {this.state.rightSide}
                   
            </Menu.Item>
        </Menu>
        );
    }
    }

  