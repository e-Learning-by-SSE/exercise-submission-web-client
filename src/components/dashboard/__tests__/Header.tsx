import renderer from "react-test-renderer";
import Header from "../Header";
import { BrowserRouter as Router } from "react-router-dom";

jest.mock("../../../services/DataService");

describe ("render Header", () => {

    it("with darkmode", () => {

        const header = renderer.create( <Router><Header onChangeVisbility={() =>{} }
         onChangeDarkMode={(darkmode: boolean) =>{} } darkmode={true} /></Router>).toJSON();

        expect(header).toMatchSnapshot();
    });
    it("without darkmode", () => {

        const header = renderer.create(<Router><Header onChangeVisbility={() =>{} }
         onChangeDarkMode={(darkmode: boolean) =>{} } darkmode={false} /></Router>).toJSON();

        expect(header).toMatchSnapshot();
    });

    it("with login", () => {

        const header = renderer.create(<Router><Header onChangeVisbility={() =>{} }
        onChangeDarkMode={(darkmode: boolean) =>{} } darkmode={false} /></Router>).toJSON();

        expect(header).toMatchSnapshot();
    });
    it("without login", () => {

        const header = renderer.create(<Router><Header onChangeVisbility={() =>{} }
        onChangeDarkMode={(darkmode: boolean) =>{} } darkmode={false} /></Router>).toJSON();

        expect(header).toMatchSnapshot();
    });

});