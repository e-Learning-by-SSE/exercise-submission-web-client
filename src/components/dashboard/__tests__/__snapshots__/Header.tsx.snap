// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Header with darkmode 1`] = `
<div
  className="ui inverted secondary menu"
>
  <div
    className="item"
    onClick={[Function]}
  >
    <button
      className="ui icon button"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="bars inverted icon"
        onClick={[Function]}
      />
    </button>
  </div>
  <div
    className="right item"
    onClick={[Function]}
  >
    <div
      className="ui checked toggle checkbox"
      onChange={[Function]}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      <input
        checked={true}
        className="hidden"
        readOnly={true}
        tabIndex={0}
        type="checkbox"
      />
      <div
        className="ui label"
        onClick={[Function]}
      >
        Darkmode
         
        <i
          aria-hidden="true"
          className="moon inverted icon"
          onClick={[Function]}
        />
      </div>
    </div>
    <a
      className="ui button"
      href="/login"
      onClick={[Function]}
      role="button"
    >
      Login
    </a>
  </div>
</div>
`;

exports[`render Header with login 1`] = `
<div
  className="ui secondary menu"
>
  <div
    className="item"
    onClick={[Function]}
  >
    <button
      className="ui icon button"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="bars icon"
        onClick={[Function]}
      />
    </button>
  </div>
  <div
    className="right item"
    onClick={[Function]}
  >
    <div
      className="ui toggle checkbox"
      onChange={[Function]}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      <input
        checked={false}
        className="hidden"
        readOnly={true}
        tabIndex={0}
        type="checkbox"
      />
      <div
        className="ui label"
        onClick={[Function]}
      >
        Normalmode
         
        <i
          aria-hidden="true"
          className="sun icon"
          onClick={[Function]}
        />
      </div>
    </div>
    <a
      className="ui button"
      href="/login"
      onClick={[Function]}
      role="button"
    >
      Login
    </a>
  </div>
</div>
`;

exports[`render Header without darkmode 1`] = `
<div
  className="ui secondary menu"
>
  <div
    className="item"
    onClick={[Function]}
  >
    <button
      className="ui icon button"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="bars icon"
        onClick={[Function]}
      />
    </button>
  </div>
  <div
    className="right item"
    onClick={[Function]}
  >
    <div
      className="ui toggle checkbox"
      onChange={[Function]}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      <input
        checked={false}
        className="hidden"
        readOnly={true}
        tabIndex={0}
        type="checkbox"
      />
      <div
        className="ui label"
        onClick={[Function]}
      >
        Normalmode
         
        <i
          aria-hidden="true"
          className="sun icon"
          onClick={[Function]}
        />
      </div>
    </div>
    <a
      className="ui button"
      href="/login"
      onClick={[Function]}
      role="button"
    >
      Login
    </a>
  </div>
</div>
`;

exports[`render Header without login 1`] = `
<div
  className="ui secondary menu"
>
  <div
    className="item"
    onClick={[Function]}
  >
    <button
      className="ui icon button"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="bars icon"
        onClick={[Function]}
      />
    </button>
  </div>
  <div
    className="right item"
    onClick={[Function]}
  >
    <div
      className="ui toggle checkbox"
      onChange={[Function]}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      <input
        checked={false}
        className="hidden"
        readOnly={true}
        tabIndex={0}
        type="checkbox"
      />
      <div
        className="ui label"
        onClick={[Function]}
      >
        Normalmode
         
        <i
          aria-hidden="true"
          className="sun icon"
          onClick={[Function]}
        />
      </div>
    </div>
    <a
      className="ui button"
      href="/login"
      onClick={[Function]}
      role="button"
    >
      Login
    </a>
  </div>
</div>
`;
